src/backends/policy_engine/souffle/souffle_policy_checker.cc

```diff
@@ -18,25 +18,59 @@
 
 #include "souffle/SouffleInterface.h"
 #include "src/backends/policy_engine/policy.h"
+#include "src/backends/policy_engine/souffle/datalog_lowering_visitor.h"
+#include "src/common/utils/filesystem.h"
 #include "src/ir/module.h"
 
 namespace raksha::backends::policy_engine {
 
+// Disambiguate "souffle" and "::souffle" with using aliases.
+// These are from the Souffle project and have a top-level namespace of
+// "::souffle".
+using Relation = ::souffle::Relation;
+using SouffleProgram = ::souffle::SouffleProgram;
+using ProgramFactory = ::souffle::ProgramFactory;
+
+// These are from Raksha's "souffle" subpackage.
+using DatalogLoweringVisitor = souffle::DatalogLoweringVisitor;
+
 static constexpr char kDatalogPolicyEngine[] = "datalog_policy_verifier_cxx";
 static constexpr char kViolatesPolicyRelation[] = "violatesPolicy";
 
-bool SoufflePolicyChecker::IsModulePolicyCompliant(const ir::Module& module,
-                                                   const Policy& policy) const {
-  std::unique_ptr<souffle::SouffleProgram> program(CHECK_NOTNULL(
-      souffle::ProgramFactory::newInstance(std::string(kDatalogPolicyEngine))));
+static bool IsModulePolicyCompliantInner(
+    const ir::Module& module, const Policy& policy,
+    std::filesystem::path facts_directory) {
+  absl::Status output_policy_status =
+      policy.OutputAsDlFactsToDirectory(facts_directory);
+  CHECK(output_policy_status.ok())
+      << "Unexpected error while outputting policy: " << output_policy_status;
```

Comment:
Why don't we just use `WriteFactsStringToFactsFile` to dump the `isSqlPolicyRule` string passed in through the swig interface?

Reply:
I don't 100% understand this comment. I've refactored things a bit so that the `isSqlPolicyRule` facts are a field on the `Policy` object and we are using `WriteFactsStringToFactsFile` to put them into a `.facts` file in the facts directory. Does that satisfy this comment?